pkg/pool-utils/contracts/BasePool.sol

```diff
@@ -186,6 +200,25 @@ abstract contract BasePool is IBasePool, BasePoolAuthorization, BalancerPoolToke
         return _MAX_SWAP_FEE_PERCENTAGE;
     }
 
+    /**
+     * @notice Returns whether the pool is in Recovery Mode.
+     */
+    function inRecoveryMode() public view override returns (bool) {
+        return _miscData.decodeBool(_RECOVERY_MODE_BIT_OFFSET);
+    }
+
+    /**
+     * @dev Sets the recoveryMode state, and emits the corresponding event.
+     *
+     * No complex code or external calls that could fail should be placed here, which could jeopardize
+     * the ability to enable and disable Recovery Mode.
+     */
+    function _setRecoveryMode(bool enabled) internal virtual override {
+        _miscData = _miscData.insertBool(enabled, _RECOVERY_MODE_BIT_OFFSET);
+
+        emit RecoveryModeStateChanged(enabled);
```

Comment:
Bit unfortunate that we have to emit this event here - ideally recovery mode would emit it and we'd just deal with storage. But this is fine for now.

Reply:
Well, I actually had it that way. This routine called super into RecoveryMode, which only emitted the event. But that seemed awkward too,  since then if derived classes didn't call super, it would. change the storage without the event, or vice versa. So I put them together to make sure both happened. A bit messy in different ways either way.

Reply:
Yeah an alternative would be something like

```
fn _privateSetRecovery() private {
  _setRecovery();
  emit event
}

fn _setRecovery() internal virtual;
```

But then if you called `_setRecovery` directly that'd be a mistake.